### PR TITLE
[all] Fix invalid Memory access in MechanicalObject and Compliant_test

### PR DESCRIPTION
**Pull request by damienmarchal**
______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by epernod** [comment]
[ci-build][with-all-tests]

File: SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/MechanicalObject.inl [hunk 1]
```diff
@@ -778,9 +778,10 @@ void MechanicalObject<DataTypes>::applyScale(const SReal sx,const SReal sy,const
     const sofa::defaulttype::Vec<3,Real> s((Real)sx, (Real)sy, (Real)sz);
     for (unsigned int i=0; i<x_wA.size(); i++)
     {
-        x_wA[i][0] = x_wA[i][0] * s[0];
-        x_wA[i][1] = x_wA[i][1] * s[1];
-        x_wA[i][2] = x_wA[i][2] * s[2];
+        for (unsigned int j=0; j<DataTypes::spatial_dimensions; ++j)
+        {
+            //x_wA[i][j] = x_wA[i][j] * s[j];
```
Comment (epernod) [inline SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/MechanicalObject.inl]: I'm quite sure there is a problem here

File: applications/plugins/Compliant/Compliant_test/Assembly_test.cpp [hunk 2]
```diff
@@ -133,13 +133,18 @@ struct Assembly_test : public CompliantSolver_test
...
-        forces.resize(2);
-        // pull the left-hand particle to the left
-        (*indices)[0]= 0; forces[0]= Vec3(-1,0,0);
-        // pull the right-hand particle to the right
-        (*indices)[1]= n-1; forces[1]= Vec3(1,0,0);
+
+        /// It is mandatory to keep this scope otherwise the WriteAccessor will hold a
+        /// pointer to the ff that will be destroyed.
+        {
+            helper::WriteAccessor< Data<helper::vector<Vec3> > > forces( ff->d_forces );
```
Comment (epernod) [inline applications/plugins/Compliant/Compliant_test/Assembly_test.cpp:140]: I don't understand, if this WriteAccessor is not used later in the code (after the scope you just added) when/why would he access the point he is holding?

File: applications/plugins/Compliant/Compliant_test/Assembly_test.cpp [hunk 3]
```diff
@@ -414,8 +409,8 @@ struct Assembly_test : public CompliantSolver_test
         // The solver
         complianceSolver = addNew<OdeSolver>(solverObject);
         complianceSolver->storeDynamicsSolution(true);
-//        complianceSolver->f_printLog.setValue(true);
-//        complianceSolver->debug.setValue(true);
+        //        complianceSolver->f_printLog.setValue(true);
```
Comment (epernod) [inline applications/plugins/Compliant/Compliant_test/Assembly_test.cpp]: I would say either removing the commented lines or removing this indent changes.

Comment (damienmarchal) [reply applications/plugins/Compliant/Compliant_test/Assembly_test.cpp]: Auto-align fault :)

File: applications/plugins/Compliant/Compliant_test/Assembly_test.cpp (hunk 2 above)
Comment (damienmarchal) [reply applications/plugins/Compliant/Compliant_test/Assembly_test.cpp:140]: I think the destructor of the accessors are called at the end of global scope and if they are referencing an object that has already been detroyed..then there is troubles.

**Comment by fredroy** [comment]
👍👍👍👍👍

**Comment by damienmarchal** [comment]
Feel free to merge, we are short in free-space on the CI :)

**Pull request merged**